File: src/clase-05/python/facade.py
from __future__ import annotations

class SubsystemA:
    def connect(self) -> str:
        return "A: conexión establecida"
    def query(self, q: str) -> str:
        return f"A: ejecutando '{q}'"
    def close(self) -> str:
        return "A: conexión cerrada"

class SubsystemB:
    def init(self) -> str:
        return "B: inicializado"
    def process(self, data: str) -> str:
        return f"B: procesando [{data}]"
    def shutdown(self) -> str:
        return "B: apagado"

# Facade: interfaz simple que orquesta los subsistemas
class SystemFacade:
    def __init__(self):
        self.a = SubsystemA()
        self.b = SubsystemB()

    def run_workflow(self, query: str) -> str:
        steps = [
            self.a.connect(),
            self.b.init(),
            self.a.query(query),
            self.b.process("resultado intermedio"),
            self.b.shutdown(),
            self.a.close(),
        ]
        return " | ".join(steps)

def main():
    facade = SystemFacade()
    print(facade.run_workflow("SELECT * FROM users"))

if __name__ == "__main__":
    main()

File: src/clase-05/typescript/facade.ts
export class SubsystemA {
  connect(): string {
    return "A: conexión establecida";
  }
  query(q: string): string {
    return `A: ejecutando '${q}'`;
  }
  close(): string {
    return "A: conexión cerrada";
  }
}

export class SubsystemB {
  init(): string {
    return "B: inicializado";
  }
  process(data: string): string {
    return `B: procesando [${data}]`;
  }
  shutdown(): string {
    return "B: apagado";
  }
}

export class SystemFacade {
  private a = new SubsystemA();
  private b = new SubsystemB();

  runWorkflow(query: string): string {
    const steps = [
      this.a.connect(),
      this.b.init(),
      this.a.query(query),
      this.b.process("resultado intermedio"),
      this.b.shutdown(),
      this.a.close(),
    ];
    return steps.join(" | ");
  }
}

if (require.main === module) {
  const facade = new SystemFacade();
  // eslint-disable-next-line no-console
  console.log(facade.runWorkflow("SELECT * FROM users"));
}
